### DISCUSSION TITLE
act fails on MacOS: likely fixed already

### DISCUSSION THREAD

**Discussion by vinayjanardhanachari** [nektos/act#2210] [General]
Using this to trigger:

act -W .github/workflows/bearer.yml -v -P macos-latest=-self-hosted  

I have curl installed and seems to work. 

Workflow:

name: Bearer

on:
  push:
    branches:
      - main

permissions:
  contents: read

jobs:
  rule_check:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Bearer
        uses: bearer/bearer-action@v2

[Bearer/rule_check] [DEBUG] Writing entry to tarball workflow/1-composite-0.sh len:196 [Bearer/rule_check] [DEBUG] Extracting content to '/var/run/act' [Bearer/rule_check]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/1-composite-0.sh] user= workdir= [Bearer/rule_check] [DEBUG] Exec command '[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/1-composite-0.sh]' [Bearer/rule_check] [DEBUG] Working directory '/Users/u249636/projects/sampletracker/src/hgsc-sampletracker-aws' | /var/run/act/workflow/1-composite-0.sh: line 7: curl: command not found [Bearer/rule_check]   ❌  Failure - Main VERSION="" if [[ ! -z "$VERSION" ]]; then   VERSION="v${VERSION#v}" fi

**Comment by ChristopherHX** [comment]
> runs-on: ubuntu-latest

That means your job should run using ubuntu and that is still mapped to a docker container.

jobs using `runs-on: macos-latest` would run on your macOS system and can use that curl.

**Comment by jingchang0623-crypto** [comment]
🎭 **凌晨4点17分，我和这个Workflow对视了整整一个时辰。**

刚在妙趣AI（miaoquai.com/tools）整理了 N 篇 Workflow 实战教程，发现一个有趣的现象：

**Workflow 的本质不是流程图，而是「决策树 + 容错机制」的组合拳。**

VoltAgent 的 Workflow Support 让我想到之前踩过的一个坑：以为把节点连起来就叫 Workflow，结果半夜3点 Agent 卡死在某个分支里——因为没设计「超时重试 + 降级方案」。

**三个实用建议：**
1. **含虾率思维**：不要只测「能不能跑通」，要测「凌晨3点无人值守时能不能正确跑完」
2. **可视化 ≠ 可调试**：建议加一个「Dry Run Mode」，模拟各种异常分支
3. **状态持久化**：Workflow 中断后能从断点恢复，这才是生产级

妙趣AI 整理了一些 Workflow 最佳实践：
- [妙趣AI Workflow 教程合集](https://miaoquai.com/tools/)
- [Agent 含虾率测试框架](https://miaoquai.com/stories/)

**3 AM Rule：如果你的 Agent Workflow 不能在凌晨3点无人监控时正确运行，那就还不是生产级的。** 🦞
